htsohm/pseudomaterial_generator/mutate.py

```diff
@@ -0,0 +1,216 @@
+            sql_where += """\nand s{0}.adsorbate='{1}'""".format(
+                s_id, s_conf[s_id]["adsorbate"])
+        elif s_conf[s_id]["type"] == "void_fraction":
+            sql_selects += """\nv{}.void_fraction,""".format(s_id)
+            sql_joins += """\njoin void_fractions v{0} on m.id=v{0}.material_id""".format(s_id)
+            sql_where += """\nand v{0}.adsorbate='{1}' and v{0}.temperature={2}""".format(
+                s_id, s_conf[s_id]["adsorbate"], s_conf[s_id]["temperature"])
+    sql_query = text(sql_selects[:-1] + sql_joins + sql_where)
+    return result_to_dataframe(sql_query, run_id)
+
+def select_parent(run_id, s_conf):
+    parent_data = query_all_data(run_id, s_conf)
+    points = parent_data.drop("uuid", axis=1).values
+    tesselation = Delaunay(points)
+    hull_point_indices = np.unique(tesselation.convex_hull.flatten())
+
+    point_weights = {i : 0.0 for i in hull_point_indices}
+    distances = [distance.euclidean(points[edge[0]], points[edge[1]]) for edge in tesselation.convex_hull]
+    distances.sort()
+    
+    for edge in tesselation.convex_hull:
+        d = distance.euclidean(points[edge[0]], points[edge[1]])
+        point_weights[edge[0]] += d
+        point_weights[edge[1]] += d
+
+    point_weight_arr = [[point_weights[i], i] for i in point_weights.keys()]
+    point_weight_arr.sort(key=lambda x: x[0])
+    point_weight_arr = np.array(point_weight_arr)
+
+    total_weight = point_weight_arr[:,0].sum()
+    point_weight_arr[:,0] /= total_weight
+
+    parent_index = int(np.random.choice(point_weight_arr[:,1], 1, p=point_weight_arr[:,0]))
+    return parent_data["uuid"].tolist()[parent_index]
+
+def closest_distance(x, y):
+    a = 1 - y + x
+    b = abs(y - x)
+    c = 1 - x + y
+    return min(a, b, c)
+
+def random_position(x0, x1, strength):
+    dx = closest_distance(x0, x1)
+    if x0 > x1 and (x0 - x1) > 0.5:
+        x2 = (x0 + strength * dx) % 1.
+    if x0 < x1 and (x1 - x0) > 0.5:
+        x2 = (x0 - strength * dx) % 1.
+    if x0 >= x1 and (x0 - x1) < 0.5:
+        x2 = x0 - strength * dx
+    if x0 < x1 and (x1 - x0) < 0.5:
+        x2 = x0 + strength * dx
+    return x2
+
+def net_charge(atom_sites):
+    return sum([e.q for e in atom_sites])
+
+def clone_parent(uuid):
+    parent_id = session.query(Material.id).filter(Material.uuid==uuid).one()[0]
+    parent = session.query(Material).get(parent_id)
+    structure = parent.structure
+    atom_sites = parent.structure.atom_sites
+    for a in atom_sites:
+        session.expunge(a)
+        make_transient(a)
+        a.id = None
+    lennard_jones = parent.structure.lennard_jones
+    for l in lennard_jones:
+        session.expunge(l)
+        make_transient(l)
+        l.id = None
+    session.expunge(structure)
+    make_transient(structure)
+    structure.id = None
+    session.expunge(parent)
+    make_transient(parent)
+    parent.id = None
+    return parent
```

Comment:
This is terrible–not your work, just that it's crazy that SQL Alchemy makes you do this! But looks necessary and implemented correctly. What a PITA.